proofs/sectorbuilder/rustsectorbuilder.go

```diff
@@ -321,25 +321,40 @@ func (sb *RustSectorBuilder) GeneratePoST(req GeneratePoSTRequest) (GeneratePoST
 
 	challengeSeedPtr := unsafe.Pointer(&(req.ChallengeSeed)[0])
 
-	// a mutable pointer to a GeneratePoSTResponse C-struct
-	resPtr := (*C.GeneratePoSTResponse)(unsafe.Pointer(C.generate_post((*C.SectorBuilder)(sb.ptr), (*C.uint8_t)(cflattened), C.size_t(len(flattened)), (*[32]C.uint8_t)(challengeSeedPtr))))
+	// a mutable pointer to a GeneratePoStResponse C-struct
+	resPtr := (*C.GeneratePoStResponse)(unsafe.Pointer(C.generate_post((*C.SectorBuilder)(sb.ptr), (*C.uint8_t)(cflattened), C.size_t(len(flattened)), (*[32]C.uint8_t)(challengeSeedPtr))))
 	defer C.destroy_generate_post_response(resPtr)
 
 	if resPtr.status_code != 0 {
-		return GeneratePoSTResponse{}, errors.New(C.GoString(resPtr.error_msg))
+		return GeneratePoStResponse{}, errors.New(C.GoString(resPtr.error_msg))
 	}
 
-	// copy proof bytes back to Go from C
-	proofSlice := C.GoBytes(unsafe.Pointer(&resPtr.proof[0]), C.int(proofs.SnarkBytesLen))
-	var proof proofs.PoStProof
-	copy(proof[:], proofSlice)
-
-	return GeneratePoSTResponse{
-		Proof:  proof,
+	return GeneratePoStResponse{
+		Proofs: goPoStProofs(resPtr.flattened_proofs_ptr, resPtr.flattened_proofs_len),
 		Faults: goUint64s(resPtr.faults_ptr, resPtr.faults_len),
 	}, nil
 }
 
+// goPoStProofs accepts a pointer to a C-allocated byte array and a size and
+// produces a Go-managed slice of PoStProof. Note that this function copies
+// values into the Go heap from C.
+func goPoStProofs(src *C.uint8_t, size C.size_t) []proofs.PoStProof {
+	chunkSize := int(proofs.PoStBytesLen)
+
+	out := make([]proofs.PoStProof, int(size)/chunkSize)
```

Comment:
is it more safe to verify that size is an integral number of chunksizes?

Reply:
If size % PoSt proof size != 0 because of a bug in rust-fil-proofs, the miner will get slashed (and programmers will become very confused). I like your recommendation to enforce size invariant at runtime.